### ISSUE TITLE
Two suggestions or questions in one: 1) is it available for other glimmer variants than libui? 2)

### ISSUE THREAD

**Issue by rubyFeedback** [AndyObtiva/glimmer-libui-cc-graphs_and_charts#1]
Hey there Andy,

just saw your blog entry.

I have two questions or requests:

1) Will this be available to other glimmer variants too, such as glimmer-gtk? The latter may be pretty useful, e. g. we could have glimmer-gtk-widgets that could show such graphs and charts. Especially this may be useful to have for use in, say, glimmer-libui, tk, SWT, also the world wide web variant. The latter could then also be used to embed in a website perhaps. For instance, to track the change of gasoline prices per year, or some other statistical use cases, which I think would be nice to have.

2) Could you perhaps add one or two more examples that relate to reallife? Specifically I am thinking here of a SIMPLE chart
that shows how some share of a company went up and down in the last 1 year. That is, code to make this work (but only the datapoints, so ultimately you can simulate it too, that is fine. For instance the rise and fall of a stock market share for any company, and to also show the flow in glimmer).

3) Ah yes, before I forgot ... a third use case. Perhaps printing this out, or simply generating a .pdf, may be useful. Not sure how difficult this is, but I kind of envision a "generate .pdf file" button, and upon clicking it you can perhaps use prawn or hexapdf to generate a .pdf file that should ideally look like the chart. Or, if that is too difficult, to just show a 2D chart in black and white, but in a .pdf file then, so that it is easier to distribute these results too.

Don't worry too much if this is cumbersome or takes away too much time, just skip that in this case, or focus on a minimal example just for showcasing the very basics of it, e. g. when people were to have such use cases and you could guide them to the correct path. As always feel free to proceed as you see fit.

2023 is soon over so I'll probably leave this suggestion as my last one for 2023. For 2024 I have some ideas ... a glimmer specification, a bit like rack ... re-usable components, also for downstream applications that want to use glimmer-widgets but without necessarily tying the whole application into the glimmer DSL ... hmm. But I'll not spam the tracker down with too many ideas; I myself have a few things to work on, and I want to test wxruby more seriously soon, to see how viable it is.

Glimmer on! \o/

**Comment by AndyObtiva** [comment]
Isn’t the example included clear enough though? It does almost what you ask for. Just replace feature A and feature B with anything (my simulation is a software feature AB testing graph)

Otherwise, this Ruby gem got extracted from the Kuiq project for Sidekiq that you stumbled upon recently. Once that project is complete, I will announce the connection between it and the Graphs and Charts library. 

Glimmer DSL for SWT already includes a line graph example, albeit not implemented with a library: https://github.com/AndyObtiva/glimmer-dsl-swt/blob/master/docs/reference/GLIMMER_SAMPLES.md#stock-ticker

This Graphs and Charts Ruby gem is only for Glimmer DSL for LibUI though. I might reimplement for other Glimmer DSLs, but only when the need arises for a real project. 

I should eventually flesh out the documentation of this Ruby gem in more details, and I might add another example or more. Once I have done so, I will close this issue. 

I think you already provide me with too many ideas without ever building anything with Glimmer (or maybe you just don’t show me what you build, which you should). If you want to really contribute, support me by building apps with Glimmer and share them with me as that provides me with indirect feedback to improve the library, and it gives me a good sense of reward for seeing apps built with it that take advantage of maximum productivity and maintainability not available in any other library. That will get my attention a lot more than ideas. I have enough ideas as it is, so if you provide me with ideas, I will ignore them. But, having more actual apps built with Glimmer is always great. The more the merrier. You can ask me for help with building real Glimmer apps when needed (even open source apps are considered real). 

Have a Merry Christmas and Happy New Year (2024)!

**Comment by MtnBiker** [comment]
basic_bar_chart.rb works, but basic_line_chart is blank. Ruby 3.4.1

But I got line_chart_absolute (cut and paste) working, so assume minor issue with basic_bar_chart

For a line chart, I would expect x and y values to be in a hash or nested array, the rocket ship syntax seems odd to me, but I am a Ruby beginner.

Looking more at the line chart, shouldn't the first points of each line start at the same x value? They are both 2030, 12, 1.

An abscissa label might have made this easier to sort out.

I better stop now as I'm confusing myself on this small screen with three charts labeled `Basic Line Graph.` I might suggest adding some dividers on the page too.

Thank you for Glimmer and all the support.

**Comment by MtnBiker** [comment]
Hash (or is it a nested array) does work
`{
						name: 'Stock 3 hash',
						stroke: [255, 165, 0, thickness: 2],
						values: [	
							[Time.new(2030, 12, 1), 20],
							[Time.new(2030, 12, 3), 10],
							[Time.new(2030, 12, 3), 80],
							[Time.new(2030, 12, 7), 30],
							[Time.new(2030, 12, 8), 17]
						],
						x_value_format: -> (time) {time.strftime("%a %d %b %Y %T GMT")},
					},`

and the first hash plots with the x-axis accurately. But a second hash then accumulate (or something) an error.

<img width="903" alt="image" src="https://github.com/user-attachments/assets/6220669a-52ed-46e6-bb63-ef9f653e5c75" />


For completeness the whole file:
`
require 'glimmer-dsl-libui'
require 'glimmer/view/line_graph'

class LineGraphAbsolute
	include Glimmer::LibUI::Application
	
	before_body do
		@start_time = Time.now
	end
	
	body {
		window('Line Graph Absolute', 900, 330) {
			line_graph(
				width: 900,
				height: 300,
				lines: [
					{
						name: 'Stock 1',
						stroke: [163, 40, 39, thickness: 2],
						values: {
							Time.new(2030, 12, 1) => 80,
							Time.new(2030, 12, 2) => 36,
							Time.new(2030, 12, 4) => 10,
							Time.new(2030, 12, 5) => 60,
							Time.new(2030, 12, 6) => 20,
						},
						x_value_format: -> (time) {time.strftime("%a %d %b %Y %T GMT")},
					},
					{
						name: 'Stock 2',
						stroke: [47, 109, 104, thickness: 2],
						values: {
							Time.new(2030, 12, 1) => 62,
							Time.new(2030, 12, 2) => 0,
							Time.new(2030, 12, 3) => 90,
							Time.new(2030, 12, 5) => 0,
							Time.new(2030, 12, 7) => 17,
						},
						x_value_format: -> (time) {time.strftime("%a %d %b %Y %T GMT")},
					},
					{
						name: 'Stock 2 with data changes',
						stroke: [0, 255, 1, thickness: 2],
						values: {
							Time.new(2030, 12, 1) => 45,
							Time.new(2030, 12, 3) => 0,
							Time.new(2030, 12, 3) => 90,
							Time.new(2030, 12, 6) => 0,
							Time.new(2030, 12, 8) => 17,
						},
						x_value_format: -> (time) {time.strftime("%a %d %b %Y %T GMT")},
					},
					{
						name: 'Stock 4 hash',
						stroke: [255, 165, 0, thickness: 2],
						values: [	
							[Time.new(2030, 12, 1), 20],
							[Time.new(2030, 12, 3), 10],
							[Time.new(2030, 12, 3), 80],
							[Time.new(2030, 12, 7), 30],
							[Time.new(2030, 12, 8), 17]
						],
						x_value_format: -> (time) {time.strftime("%a %d %b %Y %T GMT")},
					},
					{
						name: 'Stock 2 as hash',
						stroke: [160, 32, 240, thickness: 2],
						values: [
							[Time.new(2030, 12, 1), 62],
							[Time.new(2030, 12, 2), 0],
							[Time.new(2030, 12, 3), 90],
							[Time.new(2030, 12, 5), 0],
							[Time.new(2030, 12, 7), 17]
						],
						x_value_format: -> (time) {time.strftime("%a %d %b %Y %T GMT")},
					},
				] # lines 
			) # line_graph
			# display_attributes_on_hover: true
		} # window
	} # body
end # class

LineGraphAbsolute.launch`

**Comment by AndyObtiva** [comment]
@MtnBiker Yeah, the gem is an initial effort, and I realize the API could use some improvements and simplifications. 

Using => in Ruby indicates mapping a value to another, which is what goes on in a 2D graph. But, I realize that in math, often we use arrays as tuples of [x,y] values. So, this alternate format could certainly be useful. 

Arrays could be supported in a future version. I would inform you if an update is made with support for that.

**Comment by AndyObtiva** [comment]
I just read your follow up comment (I replied after reading your first comment) and it seems to say that an array does work too. If so, I’m glad you could run the alternate code with no problem.

**Comment by AndyObtiva** [comment]
@MtnBiker I will look into the code example in your second comment when I have a chance.

**Comment by MtnBiker** [comment]
Just to be clear to anyone reading this thread, the x-axis values can be inaccurate whether array or => notation.

**Comment by AndyObtiva** [comment]
@MtnBiker inaccurate how?

Could you please answer by starting a new issue thread and copying everything you said here into there? It’s better we don’t hijack this one as it’s got unrelated content in its main body.
